### PR TITLE
WIP: Refactor

### PR DESCRIPTION
**Pull request by pongraczgabor87**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,18 +1,29 @@
...
+WORKDIR $GOPATH/src/$ROOT_PACKAGE
 
 RUN apk add --update g++
 RUN go mod tidy
-RUN cd $GOPATH/src/mysql-resources-db-go-service/ && go build main.go
+RUN go build -ldflags "-X $ROOT_PACKAGE/config.AppVersion=$APP_VERSION" main.go
 
-# This container exposes port 8080 to the outside world
-EXPOSE 8080
+FROM alpine:latest
```
Comment (artofimagination) [inline Dockerfile]: Why do we need this second "from"?

File: Dockerfile [hunk 2]
```diff
@@ -1,18 +1,29 @@
+ARG EXECUTABLE_NAME=mysql-resources-db-go-service
+
 FROM golang:1.15.2-alpine
 
-WORKDIR $GOPATH/src/mysql-resources-db-go-service
+ENV ROOT_PACKAGE=github.com/artofimagination/$EXECUTABLE_NAME
```
Comment (artofimagination) [inline Dockerfile:7]: Is it only a coding convention using "github.com/..." in the path? Or does it have any other significance?

File: dbcontrollers/resources.go [hunk 3]
```diff
@@ -93,7 +93,7 @@ func (c *MYSQLController) GetResourcesByIDs(IDs []uuid.UUID) ([]models.Resource,
 	return resources, c.DBConnector.Commit(tx)
 }
 
-func (c *MYSQLController) GetResourceByID(ID *uuid.UUID) (*models.Resource, error) {
+func (c *MYSQLController) GetResourceByID(ID uuid.UUID) (*models.Resource, error) {
```
Comment (artofimagination) [inline dbcontrollers/resources.go]: So apparently passing by pointer is slower than by value... As far as I understood it is related to garbage collection. This is opposite to C++.
I will slowly adjust the rest of the code to use values in function parameters. 
There is one thing I don't understand. Why is passing pointer faster when you have a non-pointer receiver. To help to understand my question I did a small benchmark, please have a look, and could you maybe explain, if you know the answer?
https://github.com/artofimagination/go-benchmark-examples/blob/master/tests/pass-value-and-pointer-to-function_test.go

I am testing here different ways of passing a variable to a function.

File: di/container.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+			echoEngine,
+			svc,
+			cfg.DebugPProf,
+		),
+	)
+
+	return c, nil
+}
+
+func NewValidator() (*validation.Validator, error) {
```
Comment (artofimagination) [inline di/container.go:67]: Is this to check struct tags?

File: rest/controller.go [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+		return eCtx.String(http.StatusOK, "Hi! I am a user database server!")
+	})
+
+	c.echoEngine.GET("/healthcheck", func(eCtx echo.Context) error {
+		return eCtx.String(http.StatusOK, "ok")
+	})
+
+	c.echoEngine.POST("/add-resource", func(eCtx echo.Context) error {
+		resource := &models.Resource{}
+		if err := eCtx.Bind(resource); err != nil {
```
Comment (artofimagination) [inline rest/controller.go:60]: What is the purpose of this bind? Is it binding the parameters or json map of incoming request to the resource?

File: rest/controller.go [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+			return err
+		}
+
+		if err := eCtx.Validate(resource); err != nil {
```
Comment (artofimagination) [inline rest/controller.go:64]: And this one validates, whether all the required fields are filled in and it is filled in a way that passes the validation, right?

File: rest/controller.go [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+		debugPProf: debugPProf,
+	}
+}
+
+func (c *controller) Start() {
+	if c.debugPProf {
+		runtime.SetBlockProfileRate(1)
+		runtime.SetMutexProfileFraction(5)
+
+		c.echoEngine.GET("/debug/pprof/*", echo.WrapHandler(http.HandlerFunc(pprof.Index)))
```
Comment (artofimagination) [inline rest/controller.go:43]: What are these for?

File: main.go [hunk 8]
```diff
@@ -1,18 +1,75 @@
...
+}
+
+func ContextMapper() log.ContextMapper {
+	return contextMapper{}
+}
+
 func main() {
 
-	_, err := restcontrollers.NewRESTController()
+	if err := zap.RegisterEncoder(
```
Comment (artofimagination) [inline main.go:33]: Is zap just a logger library?

File: di/container.go [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+	return validation.NewValidator(v), nil
+}
+
+func newEcho(port int, validator *validation.Validator, httpErrorHandler echo.HTTPErrorHandler) *echo.Echo {
+	e := echo.New()
+
+	e.Use(echolog.RecoveryMiddleware(log.GlobalLogger()))
+	e.HTTPErrorHandler = httpErrorHandler
+	e.Validator = validator
+	e.HideBanner = true
```
Comment (artofimagination) [inline di/container.go:110]: What is the banner?

File: di/container.go [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+	"github.com/proemergotech/log/v3/echolog"
+
+	"github.com/artofimagination/mysql-resources-db-go-service/config"
+	"github.com/artofimagination/mysql-resources-db-go-service/dbcontrollers"
+	"github.com/artofimagination/mysql-resources-db-go-service/rest"
+	"github.com/artofimagination/mysql-resources-db-go-service/service"
+	"github.com/artofimagination/mysql-resources-db-go-service/validation"
+)
+
+type Container struct {
```
Comment (artofimagination) [inline di/container.go:26]: If we happen to want to replace REST interface for inter service comms, for example with gRPC, do we only need to replace Container with a new implementation?

Like creating a Container interface that either can take a REST server or gRPC or any other inter service protocol implementation.

Of course I don't intend to keep REST as the main comm protocol in the long run, although it can be kept as an alternative implementation.

So basically the question is, is this refactor still allow an easy replacement of comm interface with other services?

File: Dockerfile (hunk 2 above)
Comment (pongraczgabor87) [reply Dockerfile:7]: Before go modules this was the only way go dep could find local libs/imports.
Nowdays, most of the libs are go mod based, but should keep this structure to make sure because old libs still looking for deps in these structure.

File: di/container.go (hunk 10 above)
Comment (pongraczgabor87) [reply di/container.go:26]: If you want to use differenc comm interfaces in the same time, then yes, it supports this too, but I'm not sure in the way you think :D
I you would like to "choose/switch" between implementations you still need to add the other options into this container and you could start the proper interface within the `initialization` package. Where the container is initialized and the proper comm interface is started.

File: di/container.go (hunk 4 above)
Comment (pongraczgabor87) [reply di/container.go:67]: Create the validator for the checks. Here you could register unique validations too.
This validator is used in the echo init, so from that point echo will use this for struct/request validation.

File: di/container.go (hunk 9 above)
Comment (pongraczgabor87) [reply di/container.go:110]: Disable ascii art banner of echo. So in the service startup it will no longer appears.
https://echo.labstack.com/images/terminal.png

File: main.go (hunk 8 above)
Comment (pongraczgabor87) [reply main.go:33]: https://github.com/uber-go/zap/blob/master/README.md

File: rest/controller.go (hunk 7 above)
Comment (pongraczgabor87) [reply rest/controller.go:43]: https://www.freecodecamp.org/news/how-i-investigated-memory-leaks-in-go-using-pprof-on-a-large-codebase-4bec4325e192/

File: rest/controller.go (hunk 5 above)
Comment (pongraczgabor87) [reply rest/controller.go:60]: Is it binding the parameters or json map of incoming request to the resource?
Yes. If you dive into the code of this method, you will see this could bind json, query, url_params to your destination struct. Also handles query param list (e.g. ids) and other stuff.

File: rest/controller.go (hunk 6 above)
Comment (pongraczgabor87) [reply rest/controller.go:64]: Yeap.

File: Dockerfile (hunk 1 above)
Comment (pongraczgabor87) [reply Dockerfile]: Basicly, you could use a docker image for build your code and move the compiled executor to a smaller image for execution. With this you could reduce the size of your container image.

File: dbcontrollers/resources.go (hunk 3 above)
Comment (pongraczgabor87) [reply dbcontrollers/resources.go]: I can't see the difference you talking about.
image

File: Dockerfile (hunk 2 above)
Comment (artofimagination) [reply Dockerfile:7]: Do you mean, that possibly I can include in the future an old library that could not work, because I am using a path without the "github.com/..." in it?

File: dbcontrollers/resources.go (hunk 3 above)
Comment (artofimagination) [reply dbcontrollers/resources.go]: Haha interesting. I am on my windows machine, but tomorrow I will post my result to you (there is a clear difference between pointer and value). It lines up with some articles as well I've read about pointer performances.

File: di/container.go (hunk 10 above)
Comment (artofimagination) [reply di/container.go:26]: If I get to change the protocol, I still want to keep REST as an alternative, so if a developer prefers gRPC or REST, they can freely choose, since both implementations are accessible.  but yes, You pretty much answered. I would need to add only a new layer that calls the "dbcontroller" functions

File: dbcontrollers/resources.go (hunk 3 above)
Comment (artofimagination) [reply dbcontrollers/resources.go]: So as I promised the results I am seeing
```
peter@peter-GL63-8RCS:~/Git/go-benchmark-examples$ go test -v -bench=. -benchmem ./tests/
goos: linux
goarch: amd64
pkg: github.com/artofimagination/go-benchmark-examples/tests
BenchmarkPassValue-12                     	1000000000	         0.254 ns/op	       0 B/op	       0 allocs/op
BenchmarkPassPointer-12                   	1000000000	         0.515 ns/op	       0 B/op	       0 allocs/op
BenchmarkPassValueValueReceiver-12        	1000000000	         0.252 ns/op	       0 B/op	       0 allocs/op
BenchmarkPassPointerValueReceiver-12      	1000000000	         0.257 ns/op	       0 B/op	       0 allocs/op
BenchmarkPassValuePointerReceiver-12      	1000000000	         0.256 ns/op	       0 B/op	       0 allocs/op
BenchmarkPassPointerPointerReceiver-12    	1000000000	         0.496 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/artofimagination/go-benchmark-examples/tests	2.260s
```
I am now curious why you had different results...

File: storage/functions.go [hunk 11]
```diff
@@ -16,10 +17,9 @@ var ErrResourcesMissing = errors.New("This resources is missing or old value is
 
 const addResourceQuery = `
 	INSERT INTO 
-	resources 
-	(id, category, content) 
+	resources(id, category, content) 
 	VALUES 
-	(UUID_TO_BIN(?), ?, ?)
+	(UUID_TO_BIN(?), ?, CAST(CONVERT(? USING utf8) AS JSON))
```
Comment (artofimagination) [inline storage/functions.go:20]: Why do we need this casting and converting?

File: service/resources.go [hunk 12]
```diff
@@ -13,21 +14,19 @@ import (
...
-	log.Debug(ctx, "Add resource")
+func (s *Service) AddResource(ctx context.Context, resource *models.Resource) (*models.Resource, error) {
 	// Execute function
 	if err := s.mySQLStorage.AddResource(resource); err != nil {
-		return myerrors.WithFields(errors.Wrap(err, "mysql error"), models.HTTPCode, http.StatusInternalServerError)
+		return nil, myerrors.WithFields(errors.Wrap(err, "mysql error"), models.HTTPCode, http.StatusInternalServerError)
 	}
 
-	return nil
+	return resource, nil
```
Comment (artofimagination) [inline service/resources.go:23]: Why did you change this to return the new resource?

File: storage/functions.go (hunk 11 above)
Comment (pongraczgabor87) [reply storage/functions.go:20]: To resolve `cannot create a JSON value from a string with CHARACTER SET 'binary'`
see: https://stackoverflow.com/questions/38078119/mysql-5-7-12-import-cannot-create-a-json-value-from-a-string-with-character-set

File: service/resources.go (hunk 12 above)
Comment (pongraczgabor87) [reply service/resources.go:23]: To be able to check the db really saved the object you would like to save OR if the db creates a uuid for the object during save action you could add it to the resource itself and return with the response.

**Review by artofimagination — APPROVED** [review]
